https://www.loom.com/share/f183ff54a6ad4eebb0b8cabbe1e91864

In [1]:
!pip3 install tweepy

In [2]:
!pip install mysql-connector-python

In [2]:
consumer_key = 'CONSUMER_KEY' 
consumer_secret = 'CONSUMER_SECRET'
access_token = 'ACCESS_TOKEN'
access_token_secret = 'ACCESS_TOKEN_SECRET'
bearer_token = 'BEARER_TOKEN'

In [6]:
import tweepy
import time 
import pandas as pd 
from sqlalchemy import create_engine

engine = create_engine('mysql+mysqlconnector://admin:sql_2021@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter?charset=utf8')
client = tweepy.Client(bearer_token=bearer_token, consumer_key=consumer_key, consumer_secret=consumer_secret, access_token=access_token, access_token_secret=access_token_secret, return_type=dict, wait_on_rate_limit=True)
u_id = client.get_user(username='Wayfair')['data']['id']

data = { 'username': list(), 'following': list() }

followers = client.get_users_followers(id=u_id,max_results=10)
for f in followers['data']:
    this_name, this_id = f['username'], f['id']
    following = client.get_users_following(id=this_id,max_results=50)
    print(following)
    if 'data' in following:
        for ff in following['data']:
            data['username'].append(this_name)
            data['following'].append(ff['username'])
df = pd.DataFrame(data)
print(df)
df.to_sql('AccountsFollowed', engine, if_exists='replace', index=False)

{'data': [{'id': '16101626', 'name': 'Wayfair.com', 'username': 'Wayfair'}, {'id': '168216285', 'name': 'Alyse Eady FOX 5', 'username': 'alyseeadyFOX5'}, {'id': '3298912237', 'name': 'Natalie Whittingham Burrell 🐥', 'username': 'natlawyerchic'}, {'id': '1367694976881147905', 'name': '🟣 George', 'username': 'OneGSisneros'}, {'id': '1433345786469826561', 'name': 'Sisa', 'username': 'TheTitanBaddie'}, {'id': '1890107336', 'name': 'PTZtv', 'username': 'PTZtv'}, {'id': '16347506', 'name': 'Atlanta Falcons', 'username': 'AtlantaFalcons'}, {'id': '121546530', 'name': 'Everything Georgia', 'username': 'GAFollowers'}, {'id': '21094888', 'name': 'WSB-TV', 'username': 'wsbtv'}], 'meta': {'result_count': 9}}
{'data': [{'id': '16101626', 'name': 'Wayfair.com', 'username': 'Wayfair'}, {'id': '85307509', 'name': 'Crystal Palace F.C.', 'username': 'CPFC'}, {'id': '1446646534443966480', 'name': 'Amy Schneider', 'username': 'Jeopardamy'}, {'id': '27042513', 'name': 'Hugh Jackman', 'username': 'RealHughJ

{'data': [{'id': '1179342376088694784', 'name': 'คู่แท้อยากหาประสบการณ์', 'username': 'Joey27943656'}, {'id': '1402667805472813058', 'name': 'หาคนเลียหี', 'username': 'wonder4289'}, {'id': '1420063752003624969', 'name': '48++50.เจ้เองคะ', 'username': 'L6NaQByW1kvTpo5'}, {'id': '1222496188877795328', 'name': 'ติ๊ดตี่ จินดาวงษ์', 'username': 'BLDUM51mg3W88rX'}, {'id': '1423517290222346243', 'name': 'สาวใหญ่40 อัพรับงาน', 'username': 'ou0816213413'}, {'id': '1339544263810433024', 'name': 'แอนค่ะ', 'username': 'RvVIFOCTXcvVQsx'}, {'id': '1350278255329243140', 'name': 'Sa', 'username': 'Sa70853649'}, {'id': '1371082939849576459', 'name': 'คู่สูงวัย', 'username': 'Nnnnn32108449'}, {'id': '1440949575053807621', 'name': 'XO', 'username': 'CXIIICP'}, {'id': '1351067104489959425', 'name': 'แจ๊ะแจ๊ะพาเพลิน', 'username': '1n91IFkWUGI9o40'}, {'id': '780538188813127680', 'name': 'ชอบแอบแม่ แม่เพื่อน สาวใหญ่ (Family immoral sex )', 'username': 'Kunsonthayaa'}, {'id': '1050683866665480193', 'name': 'ช

{'data': [{'id': '1222650745901633536', 'name': 'City of Chicago', 'username': 'chicago'}, {'id': '357286827', 'name': 'Elgin, Illinois', 'username': 'Elgin_IL'}, {'id': '2425151', 'name': 'Meta', 'username': 'Meta'}, {'id': '180505807', 'name': 'Instagram', 'username': 'instagram'}, {'id': '372817561', 'name': 'Bed Bath & Beyond', 'username': 'BedBathBeyond'}, {'id': '17498138', 'name': 'Overstock', 'username': 'Overstock'}, {'id': '61349667', 'name': 'Cymax', 'username': 'cymax'}, {'id': '17448575', 'name': 'JCPenney', 'username': 'jcpenney'}, {'id': '19336500', 'name': "Lowe's", 'username': 'Lowes'}, {'id': '14791918', 'name': 'The Home Depot', 'username': 'HomeDepot'}, {'id': '432531569', 'name': 'Sauder Furniture', 'username': 'SauderUSA'}, {'id': '89084561', 'name': 'Target', 'username': 'Target'}, {'id': '148464806', 'name': 'Ashley', 'username': 'AshleyHomeStore'}, {'id': '18033198', 'name': 'Hayneedle', 'username': 'hayneedle'}, {'id': '39820738', 'name': 'Walmart Inc.', 'user

In [27]:
engine = create_engine('mysql+mysqlconnector://admin:sql_2021@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter?charset=utf8')
client = tweepy.Client(bearer_token=bearer_token, consumer_key=consumer_key, consumer_secret=consumer_secret, access_token=access_token, access_token_secret=access_token_secret, return_type=dict, wait_on_rate_limit=True)
u_id = client.get_user(username='Wayfair')['data']['id']

data = { 'username': list()}

followers = client.get_users_followers(id=u_id,max_results=150)
for f in followers['data']:
    this_name = f['username']
    data['username'].append(this_name)
print(data)
         
df = pd.DataFrame(data)
print(df)
df.to_sql('WayfairFollowers', engine, if_exists='replace', index=False)

{'username': ['CainC_xD', 'KEWZIIE', 'Tantann1', 'erniebets', 'TracieR66729489', 'mkmorani', 'MolliesophiaLtd', 'Louise_cole1', 'xhaopengfei', 'riaction20', 'Arsalan36933021', 'martybeaux', 'ConvenConcepts', 'JohnLGiovanni', 'j_lisaj36946344', 'johnny4896', 'Maurelia7604', 'notastringbean', 'fencifang', 'Mia123syy', 'DeliaFiebich', 'castusmom', 'sexyalexislove3', 'Meredit50334575', 'queenpink71', 'SUPERPIMPNIGGA', 'nSearchofWonder', 'CathyAnn1355', 'disert_rose', 'SimplyAubs', 'Terri06707145W', 'Mokgaotsi_', 'BerkshireSTEM', 'kthaw66', 'JimmyBessing', 'JeopardyJP', 'chicago_bullish', 'emaloney84', 'ZonaGirl13', 'C_Reed214SG', 'waltedman', 'habitabulous', 'Jasebuggy', 'PWAdvocacy', 'RajendranHarsha', 'twoguysandavan', 'ODfz77J5xEa4vtl', 'AdhilM', 'carrieinrancho', 'WaterTableLamp', 'RRasmor', 'loulou34f', 'Chillphil06', 'JanetNe73624057', 'Jco3488', 'leojames0022', 'rajat_8', 'JohnYanni1970', 'faerie32982', 'de761761', 'OhsnapitsTK1', 'Giselle_Reds', 'LynRikkiS', 'sonjabuckle', 'sealalp

In [1]:
%load_ext sql

In [3]:
%sql mysql://USERNAME:PASSWORD@HOST/DATABASE

# Exploratory queries
## 1.)
This query was performed make sure that there were not any null values inserted into the data. 
Null values could have meant there was an error either with the data collection or putting it in mysql which
would may have meant our data was not correct

In [7]:
%%sql 
SELECT *
FROM AccountsFollowed
WHERE following IS NULL;


 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
0 rows affected.


[]

From this query I learned that that were no null values inserted into the data meaning that we can move on and not reexamine our data collection

## 2.)
This query is done to see how many different users we have selected data from on what accounts they follow in addition to wayfair. This is done to insure that we are selected the proper amount of data from all users

In [20]:
%%sql 
SELECT COUNT(DISTINCT(username))
FROM AccountsFollowed;


 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
1 rows affected.


[(9,)]

This query shows that only 9 different users are in our dataset which will signfigantly affect the conclusions we can draw. 
If I was able to get the code working correctley it would have been much more users that I gathered data from. 

## 3.) 
This query was done to see the number of accounts following I have been able to get data where it is grouped by each 
Wayfair follower. This is done to verify if the amount of followers we are getting from each account makes sense

In [11]:
%%sql
SELECT 
    username, 
    COUNT(*) AS number_of_following
FROM AccountsFollowed
GROUP BY username;


 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
9 rows affected.


[('TracieR66729489', 9),
 ('mkmorani', 50),
 ('MolliesophiaLtd', 8),
 ('Louise_cole1', 50),
 ('xhaopengfei', 50),
 ('riaction20', 43),
 ('Arsalan36933021', 50),
 ('martybeaux', 50),
 ('ConvenConcepts', 20)]

The results show that there are a few accounts with far less accounts following than the others 
because due to rate limits I capped the amount of accounts following I am checking for each user at 50. This is likely due to the high variation of accounts people follow where some users follow thousands and others follow very few accounts

## 4.)
This query is seeing how many different accounts that people who follow wayfair are following that are in our dataset. THis shows wether or not we have a large enough number in the datset to draw good conclusions from

In [67]:
%%sql 
SELECT count(*) AS accounts_followed_count
FROM AccountsFollowed

 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
1 rows affected.


[(330,)]

This query showed that there are 330 total following accounts in our dataset that we can use to draw information from to inform our advertising decision. This query shows that there is simply not enough data and if I had figured out how to, ideally there would have been more. 

## 5.) 
In this query I am performing here is to checking on the wayfair followers table. I made a request to get the username of 150 wayfair followers and put them in a seperate table so the count should be 150. I did this to have an additional table where we could store more information in the future on each of these followers that follow wayfair. 

In [79]:
%%sql 

SELECT COUNT(DISTINCT(username)) AS count_of_users
FROM WayfairFollowers;


 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
1 rows affected.


[(150,)]

From this query we can see that 150 wayfair followers we succesfuly put into the sql table menaing we can move forward from the data collection process

# Primary Question
My primary question is to find what are the most frequentley followed accounts by people that follow wayfair. In my project overview I am targeting a high value wayfair customer so this shows what accounts those high value customers also follow. This query is creating a view and joining the accounts that in the wayfair followers into the the accounts followed table. Thiw will make further queries easier and eliminate the need for joins

In [53]:
%%sql 
CREATE OR REPLACE VIEW accounts_following AS
    SELECT
        af.username,
        following
    FROM AccountsFollowed af
    LEFT JOIN WayfairFollowers wf
        ON af.username = wf.username;

 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
0 rows affected.


[]

There is not any information gained from this query, it is simply done to make the next two queries easier. 

# Secondary Queries
## 1.)
Relating to my primary question, I wanted to figure out what other accounts are followed more than once by twitter users. This query was performed to get the accounts that wayfair followers are following with counts greater than 1. I did greater than 1 becuase the dataset I am working with is so small. This will show what other areas of interest a high potential wayfair customer has and where wayfair should target their advertising

In [60]:
%%sql
WITH rank_cte AS (
    Select 
        following, 
        count(*) AS follow_count, 
        RANK() OVER(
            ORDER BY COUNT(following) DESC
    ) AS account_rank
    FROM accounts_following
    GROUP BY following
    ORDER BY count(*) DESC
)
Select 
    following, 
    follow_count,
    account_rank
FROM rank_cte
WHERE follow_count > 1


 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
2 rows affected.


[('Wayfair', 8, 1), ('BBCNews', 2, 2)]

This query showed that besides wayfair, the only other account the two people follow among the 9 users is BBCnews, which even for a small dataset seems low. This shows that maybe wayfair should have ads on bbcnews or on their twitter page at least but we need a larger dataset for this query to be more beneficial

## 2.)
For my last query I wanted to find the overall percentage of commonly followed twitter accounts are from all the accounts that are followed. This secondary query is digging deeper into seeing what percentage of wayfair followers follow unqiue twitter accounts. This shows if the data will be useful to find commonalities and exploit those commonly followed twitter accounts for advertising tareting high potential wayfair customer.

In [65]:
%%sql
SELECT 
    ROUND(COUNT(DISTINCT(following))/COUNT(following) * 100, 2) AS unique_following_percent
FROM accounts_following;

 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
1 rows affected.


[(Decimal('97.58'),)]

The results of this show that most wayfair followers following unqiue twitter accounts so it is going to be hard to find commonly followed accounts in this dataset. This shows we are going to need to gather a lot of data to find commonly followed twitter accounts, something that I was not able to do given my expertise on gathering data

In [80]:
!pip install nbmerge

  Created wheel for nbmerge: filename=nbmerge-0.0.4-py2.py3-none-any.whl size=6413 sha256=26ed48ba57c7b64eb8be956a80ca1dd185c487214d027aca6cefd9c014f79aa5
  Stored in directory: c:\users\colem\appdata\local\pip\cache\wheels\0b\a8\31\4ae563f674b488c82ee72cfe0f86487c21532db97ae910b264
Successfully built nbmerge


In [84]:
!nbmerge TwitterAPIcollection.ipynb sql_analysis_cole_montoya.ipynb -o presentation.ipynb

Traceback (most recent call last):
  File "C:\Users\colem\anaconda3\envs\spring_2022\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\colem\anaconda3\envs\spring_2022\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\colem\anaconda3\envs\spring_2022\Scripts\nbmerge.exe\__main__.py", line 7, in <module>
  File "C:\Users\colem\anaconda3\envs\spring_2022\lib\site-packages\nbmerge\__init__.py", line 202, in main
    write_notebook(nb, sys.stdout)
  File "C:\Users\colem\anaconda3\envs\spring_2022\lib\site-packages\nbformat\__init__.py", line 169, in write
    fp.write(s)
  File "C:\Users\colem\anaconda3\envs\spring_2022\lib\encodings\cp1252.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
UnicodeEncodeError: 'charmap' codec can't encode character '\U0001f425' in position 2804: character maps to <undefined>


In [1]:
!pip install nbmerge

In [4]:
!jupyter nbconvert presentation.ipynb --to slides --post serve

[NbConvertApp] Converting notebook presentation.ipynb to slides
[NbConvertApp] Writing 655562 bytes to presentation.slides.html
[NbConvertApp] Redirecting reveal.js requests to https://cdnjs.cloudflare.com/ajax/libs/reveal.js/3.5.0
Traceback (most recent call last):
  File "C:\Users\colem\anaconda3\envs\spring_2022\Scripts\jupyter-nbconvert-script.py", line 10, in <module>
    sys.exit(main())
  File "C:\Users\colem\anaconda3\envs\spring_2022\lib\site-packages\jupyter_core\application.py", line 264, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "C:\Users\colem\anaconda3\envs\spring_2022\lib\site-packages\traitlets\config\application.py", line 846, in launch_instance
    app.start()
  File "C:\Users\colem\anaconda3\envs\spring_2022\lib\site-packages\nbconvert\nbconvertapp.py", line 346, in start
    self.convert_notebooks()
  File "C:\Users\colem\anaconda3\envs\spring_2022\lib\site-packages\nbconvert\nbconvertapp.py", line 518, in conve

In [6]:
!pip install rise

In [7]:
import rise